### DISCUSSION TITLE
Can Microsoft's AutoGen run ACE?

### DISCUSSION THREAD

**Discussion by ippolite** [daveshap/ACE_Framework#59] [General]
Hello,
Has anyone tried the Microsoft's [AutoGen thing](https://github.com/microsoft/autogen)?

From the looks of it, we can assign a separate agent for each layer and kinda get at least one bus for free.
I don't quite understand how the communication between agents goes, so I would appreciate help in determining how this is different from ACE's busses.

(Also, I'd try plugging a local-online model mix in there, but not in the coming few days)

**Comment by hexadecimal-REBOOT** [comment]
The primary difference between AutoGen and ACE (Autonomous Cognitive Entity) seems to be in their focus and capabilities. AutoGen is designed for multi-agent conversations and provides a rich set of features geared toward conversational workflows, performance tuning, and a drop-in replacement for various inference APIs. It aims to simplify complex language learning model (LLM) workflows by employing a multi-agent system that can interact with each other and humans. 

On the other hand, ACE is more of a middleware that provides a framework for networked applications, focusing on cross-platform operability and real-time communication. It doesn't inherently provide a multi-agent conversational model, although it could potentially be adapted to support such a feature.

As for buses, it seems like AutoGen abstracts much of the underlying communication between agents, likely handling it internally. In ACE, buses (communication channels) are more explicitly managed.

Integrating a local-online model mix in AutoGen would likely depend on how well it can interact with external systems and data sources, something that ACE is explicitly designed to handle.

So, while both are frameworks, they serve different needs and offer different features.

**Reply by ippolite** [reply]
Thank you for the reply.

I think I should have given more context. I've seen [this single video](https://www.youtube.com/watch?v=JZduiYGYRJ0) of the implementation of AutoGen (it's in Russian, but auto-generated captions get the jist of it). 

From what I've seen, there was several autonomous agents that chatted. And I mean in natural English language. It's just I was unable to tell if they chat, like, only one way around, from agent 1  -> to agent 4, or if they can also talk in the other direction (say, agent 3 to agent 1, without agent 2 knowing).

So, several entities with their own respective goals and parts work together on a user-defined task. 
When I saw that, _it just clicked_ for me. I can assign a layer to each of the agents! Isn't that thingamajig an example of out-of-the-box ACE-implementation-ready system?
It also has a page on implementing local LLMs into it. So, basically, someone just did a thing I wanted to achieve -- have something to run a mixture of local-online models so they can handle sensitive information inside one physical machine.

**Comment by samgriek** [comment]
The ACE Framework is a conceptual framework and really it is up to the implementation.  It is inherently multi-agent as each layer is in fact and agent that forms cognitive networks with other layers through the north and southbound buses.  Theoretically you could implement a limited scope ACE Framework agent on AutoGen, but not sure how that would pan out in a practical sense.
